Add support for async functions which run to completion (#1)
Matthias247/rfcs: text/0000-run-to-completion-async-fn.md

would act on memory that does not represent the actual data to transmit anymore.
This would lead to undefined behavior.

One workaround to this problem is to use owned objects within completion based
operations. E.g.

```rust
async fn transmit_data() {
    let buffer: String = "Data to transfer".into();
    let data: Bytes = buffer.into();
    let (bytes_transferred, data) = engine.send(data).await;
}
```

The downside of this API is that it is a lot less flexible than the original
slice based API. We can now only call the API with a certain owned buffer type.
We also can't use cheap [async] stack based buffers anymore, and might need to
add support for atomic reference counting. Therefore this async abstraction
layer would not represent a zero-cost abstraction anymore, compared to the
original C version of those APIs.

By introducing run-to-completion async functions, we can model the APIs exactly
as described in the original example.

## Offer protection against accidental returns

For some function it is really important that it actually runs to completion.
The reason is typically that this function forms an atomic transaction. If the

> transaction it cancelled in the middle, the objects that this transaction is

Reviewer: ```suggestion
transaction is cancelled in the middle, the objects that this transaction is
```

> manipulating end up in an invalid state.

Reviewer: ```suggestion
manipulating will end up in an invalid state.
```

this mechanism is neither performant nor necessarily safe. Blocking on an async
task in a destructor means all other async tasks in the root executor can now
no longer make progress for the duration of the `drop`. And if this mechanism
would be used inside a single-threaded executor a deadlock could occur,
since resources inside `block_on` can require the eventloop which drove the
original `async fn` to make progress - which won't happen since the thread is
blocked on waiting for the cleanup `Future`.

`async fn`s with run to completion semantics will naturally allow transactions
to run to completion - in the absence of panics. Cleanup code can just be regular
code at the end of a function, and no RAII guards are required. Therefore the
new function type will reduce complexity for such transactions.

In addition to that it will lower the chance of accidental errors. There does
not exist an implicit return path in `.await` anymore, which might have not been
handled properly.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The proposal adds 2 new items to Rust:
- A new `Future` type, which will reside in `core::task`. In the remains of this

>   document we will refer to this type as `RunToCompletionFuture`.

Reviewer: ```suggestion
  document we will refer to this type as `CompletionFuture`.
```

What about that? :)

Author: I thought about it too. Definitely more concise. I was only not sure whether it conveys that well what it is. I'm personally fine with both

Reviewer: Imo I find it says the same thing but we can always leave the bike shedding up to the real post :)

Reviewer 2: I agree that `CompletionFuture` is more concise.

macro to cancel operations/`Future`s which do not make use of
run-to-completion semantics. This would for example be async channels, timers,
mutexes, semaphores, etc.

In order to cancel run to completion operations, the `CancellationToken` must be 
forwarded up to a point where cancellation is checked. This would e.g.
require a library which e.g. utilizes `io_uring` to perform async IO operations
to take a `CancellationToken` as an argument and listen for the cancellation
request while the operation is still pending. The signature of such a method
would e.g. be

```rust
async completion fn read_with_uring(
    &mut self, buffer: &mut [u8], cancel_token: &CancellationToken
) -> Result<usize, IoError>;
```

The implementation of such a method would rather be complicated - but only needs
to occur within the runtime which provides the method. Most end user code is
simply expected to forward the `CancellationToken`:

```rust
async completion fn read_all_with_uring(
    reader: &mut Reader, buffer: &mut [u8], cancel_token: &CancellationToken
) -> Result<(), IoError> {
    let mut offset = 0;
    while offset != buffer.len() {

>         let read = reader.read_with_uring(&buffer[offset..], cancel_token)?;

Reviewer: ```suggestion
        let read = reader.read_with_uring(&buffer[offset..], cancel_token).await?;
```

        offset += read;
    }
    Ok(())
}
```

Runtimes could also decide to forward `CancellationToken`s implicitely through
task-local storage. Or the ability to forward a `CancellationToken` could be
added to the [`std::task::Context`](https://doc.rust-lang.org/1.42.0/std/task/struct.Context.html)
type in the future. Such a mechanism is however outside of the scope of this
proposal.

## Timeout handling with `async completion fn`

Another common case where a `select!` like functionality is needed besides
checking for cancellation is for timeouts. The current `select!` macro could not
be used to time-out on `async completion fn`, since `select!` does not drive
the aborted branches to completion.

**Therefore the following example is not valid**:

```rust
async completion fn read_with_timeout(
    reader: &mut Reader, buffer: &mut [u8], timeout: Duration
) -> Result<usize, IoError> {
    select! {
        read_result = reader.read_with_uring(&buffer[offset..]) => read_result

>         _ = timer.delay_for(timeout) => Err(IoError::Timeout)

Author: ```suggestion
        _ = runtime::timer::delay_for(timeout) => Err(IoError::Timeout)
```

    }
}
```

Instead of this, a new mechanism is required which guarantees that the non
timeout branches will be cooperatively cancelled and driven to completion after
the timeout occured. That mechanism can be provided by runtimes in a variety of
fashions. They could either provide a variant of the `select!` macro which is
usable for `async completion fn`. Or they could provide a dedicated timeout
method:

```rust
async completion fn read_with_timeout(
    reader: &mut Reader, buffer: &mut [u8], timeout: Duration
) -> Result<usize, IoError> {

>     runtime.with_timeout(timeout, async completion move {

Reviewer: where does runtime here come from?

Author: Good catch. I will just make it `runtime::with_timeout`. Otherwise it would need to be passed

>     runtime.with_timeout(timeout, async completion move {

Author: ```suggestion
    runtime::with_timeout(timeout, async completion move {
```

- The trait will stay object safe. We will be able to build traits on top if it -
  e.g. for IO completion based TCP streams - which still support dynamic dispatch.
- The task and waker system are completely unmodified, and will work the same
  fashion for both `Future` types.

# Drawbacks
[drawbacks]: #drawbacks

### Additional complexity through another `Future` and function type

The main drawback of adding another async function type is that it introduces
yet another concept that developers need to be aware of. It thereby increases
complexity of the language. Even if we add support for run to completion
`async fn`, developers still need to be aware about normal `async fn`, and the
fact that those behave different than normal functions and can return in
`await` points.

However even without adding this new feature, developers are expected to
experience the complexity of a variety of different behaviors in async functions.
They already need to be aware about the current cancellation behavior. They will
also need to be aware that cancelling some of the current async functions (e.g.
the proposed `scope` function for structured concurrency support in tokio) will

> will lead to `panic`s or `abort`s, since there exists not way to handle a

Reviewer: ```suggestion
will lead to `panic`s or `abort`s, since there exists no way to handle a
```

themselves. It should mainly be generated by the compiler. In addition this low
level primitives like IO completion based socket types would need to be implemented
by runtime authors - but those will need unsafe code anyway (they need to pass
raw pointers to underlying C code or the kernel for an amount of time which can
not be checked by lifetimes). Most users are purely expected to use high level
`async completion fn`s, which will not require unsafe code. 

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

The proposal presented here provides a way to add run to completion semantics
for asynchronous functions in a backwards compatible and minimally invasive way.
Since neither the way an async function desugars into a generator/state-machine nor
the way wakeups (the task system) are performed are touched, the effort to integrate
this mechanism should be manageable.

The mechanism will allow us to provide true zero cost abstractions for completion
based operations. Those already exist in other languages (C++, Zig, C#, Kotlin,
etc), but not yet in Rust.

`async completion fn` will also introduce an async function type which behaves

> more simliar to normal functions that current `async fn`s do. It might therefore

Reviewer: ```suggestion
more similar to normal functions than current `async fn`s do. It might therefore
```

have been interesting to see `async completion fn` as the new "default" async
function type. However this is not possible anymore, since `async fn` is already
stabilized.

And while the introduction of another type of async function type sounds like
as an additional complexity at first, it also has its benefits:

Users will continue to be able to write simple and safe `Future` implementations
with cancel-anytime semantics manually, and be able to use them in combination

> with powerful flow-control macros like `select!` and `join!`. A lot of types

Reviewer: The big issue I see is that this affects leaf futures and therefore would require the entire future chain to be a completion future. Which is quite unfortunate.

Reviewer: I guess if we provided some `tokio::spawn_completion` type then you could use its join handle to "forget" about that. That should be a somewhat and minimal way to not have to funnel completion up the chain.

Author: You mean and leave it running in the background?
I think in order to avoid having those runaway tasks we will implement the `scope` proposal to work based on this.

Reviewer: Basically, what I have now for uring does this, even if the future is canceled it will still run in the background. But I think that is useful to mix with other control flow.

which can be used to drive `Future`s to completion and to query a `Future`s
execution status. `Future`s can be dropped at any point in time to cancel an
ongoing async operation.

The latter ability made cancellation in async rust a lot easier to achieve than
in most other environments. However the design choice also introduced some
gaps, which the new async function type aims to fill.

## Support for completion based operations

Todays `Future`s do not allow us to support completion based operations in
a flexible and safe fashion.

**Completion based operations are operations which follow the follow sequence:**
1. We start an asynchronous request, and provide references to all necessary
  resources to the engine which executes the requests
2. We later receive a notification from the engine that the request finished.
  This notification can be delivered in a variety of ways: It could be callback
  issued on a background thread, a signal or interrupt handler,
  or it could be dequeued through a completion port/queue on an arbitrary thread.
3. We acquire the result of the operation, which finalizes the operation.

>   after this step all resources which had been borrowed for the duratino of the

Reviewer 3: ```suggestion
  after this step all resources which had been borrowed for the duration of the
```

`unsafe` here mainly in order to enforce a contract with the caller.

However we do not expect a lot of people to implement `RunToCompletionFuture`
themselves. It should mainly be generated by the compiler. In addition this low
level primitives like IO completion based socket types would need to be implemented
by runtime authors - but those will need unsafe code anyway (they need to pass
raw pointers to underlying C code or the kernel for an amount of time which can
not be checked by lifetimes). Most users are purely expected to use high level
`#[completion] async fn`s, which will not require unsafe code.

We still expect people to write low level `Future`s by hand, for example to
implement their custom channel type, timer, mutex, etc. However those types
are typically supposed to be synchronously cancellable, and therefore people would continue to implement `Future`s instead of `RunToCompletionFuture`.
the main exception here are future types which need to run to completion, like
wrappers around IO completion operations. These are however unsafe by nature
and require careful implementation and review. Therefore the `unsafe` attribute

> added here is justified.

Reviewer 2: We also need to consider combinator authors.  Combinators can be much more efficient in this design, but they will require `unsafe` code internally.

Reviewer 4: +1. Combinators and any wrapping of futures. I'd like to see a section on combinators specifically. I fear that two disjunct traits would lead to a lot of duplication in combinators.

Reviewer 4: In fact, I think that unfortunately combinators are so deeply affected that the proposal becomes infeasible in its current state:

Say a combinator like select: once returned from it, what happens to the unfinished future? Return to caller just like today? Who ensures the future is polled again before dropped?

Same for FuturesUnordered: Poll a couple of items out, but leave some inside, then drop it - compiler is none the wiser -- and the invariant is broken.

**In general**: _A combinator that accepts completion futures **must poll all of its inner futures to completion before returning from any of its async functions**. It can neither store or return any incomplete futures because then they can be arbitrarily dropped._

Unfortunately, this gives us three bad options:

- Combinators disallow this - severely degrading their utility.
- Combinators allow this, but lose most of the guarantees of this proposal
- Drop-bombs - panic at runtime if futures and combinators are not polled to completion at drop-time.

It all boils down to this existing drop-semantics we have today - whether we're in async or not. Hence, some rudimentary linear type features are necessary, as far as I can tell.

Reviewer 2: @betamos a combinator that accepts a completion future would itself be a completion future, so this problem is pushed up to the caller and ultimately to the executor.

> Same for FuturesUnordered: Poll a couple of items out, but leave some inside, then drop it - compiler is none the wiser -- and the invariant is broken.

`FuturesUnordered` is a `Stream`, and a completion variant would be needed for this as well.  One approach would be to instead buffer all of the futures into a single `Vec` before returning them.

That said, `FuturesUnordered`’s approach is an anti-pattern in many cases.  A better API is

```rust
let (result1, result2, result3) = select!(completion_future_1, completion_future_2, completion_future_3);
```

Reviewer 4: I hear you, but I think if we invalidate people's existing use of combinators, we're gonna have a harder time getting this through review. There's a lot more convincing to do.

FuturesUnordered is used under the hood in many other combinators and stream transforms too..

How would you solve the standard issue of:
- A TCP Listener produces a stream of Connections
- For each connection concurrently:
  - Run a handler which returns a result. The handler is a completion future.
- Collect the results out-of-order (in a buffer that never exceeds O(concurrent connections))

Reviewer 2: We could use a completion version of `FuturesUnordered` that itself must be polled to completion.  Of course, there would need to be a method for cancellation as well.

Author: That is basically the structured concurrency `scope/TaskGroup/Nursery` thing.

https://github.com/tokio-rs/tokio/pull/2579 had an implementation for Rust/tokio without support for run to completion tasks. https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md proposes a `TaskGroup` for Swift which is basically the same: A `FuturesUnordered` for run-to-completion coroutines with builtin support graceful cancellation.

And as said [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/run.20to.20completion.20async.20fn/near/222140636) I don't think the current `select!/join!` combinators should go away. They will still have their roles in waiting e.g. for a channel and a timer in parallel.

The new combinators will focus more on the orchestration of bigger tasks. E.g. on the server example that @betamos  had the TCP listener would open a task group, and start the processing of each connection as a subtask in this group. If the listener gets cancelled (e.g. via `SIGINT`), the subtasks also receive a graceful cancellations signal which still allows them to e.g. finish processing the current request, and then shut down.

Author: Btw: I think the original comment from @DemiMarie was about something different. I interpret it more has "combinators which can assume run to completion guarantees can be more efficient and offer different features", than a "old combinators are no longer useful".

I fully agree with that, but it's to be seen what combinators we really need. The number 1 on my list is really the `TaskGroup` thing, which however could also enable borrowing from the parent tasks as a new feature (like crossbeam scopes).

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

The proposal presented here provides a way to add run to completion semantics
for asynchronous functions in a backwards compatible and minimally invasive way.
Since neither the way an async function desugars into a generator/state-machine nor
the way wakeups (the task system) are performed are touched, the effort to integrate
this mechanism should be manageable.

The mechanism will allow us to provide true zero cost abstractions for completion
based operations. Those already exist in other languages (C++, Zig, C#, Kotlin,
etc), but not yet in Rust.

`#[completion] async fn` will also introduce an async function type which behaves
more similar to normal functions than current `async fn`s do. It might therefore
have been interesting to see `#[completion] async fn` as the new "default" async
function type - which might be the mostly used function type used by application

> writers in the end. However this is not possible anymore, since `async fn` is

Reviewer 2: It is possible in Rust 2021 :smile:

This RFC leaves it open whether there should be a type relation between
`Future` and `RunToCompletionFuture`, or whether we just need to teach the
compiler about both types, and when the `.await` of the other type is possible.

Normal `Future`s can be used whenever a `RunToCompletionFuture` is required.
In that case the `Future`s support cancellation by drop, but the caller would
not make use of the capabilty and always drive them to completion.

A blanket impl like the following might be a possibility::

```rust
impl<F, T> RunToCompletionFuture for F
where F: Future<Output=T> {
    type Output = T;

    unsafe fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        // forward the poll call to the `Future` implementation
        <Self as Future>::poll(self, cx)
    }
}
```

# Future possibilities
[future-possibilities]: #future-possibilities

The RFC outlines a few additional ideas which will be interesting to look at
in the future:

## Keyword for async run to completion function

A future Edition of Rust could adopt a different keyword for async completion
functions in order to reduce the verbosity. This should be done

> after carefully studying the usage of the various `async fn` types.

Reviewer 2: ```suggestion
after carefully studying the usage of the various `async fn` types.

It would also be possible to make `RunToCompletionFuture` be the default.  `#[completion] async` would become just `async`, and `async` would become `#[completion(false)] async` (or similar).  This transformation can be performed automatically.
```

Author: It's totally possible. But I would rather not recommend this, since those kinds of semantic changes mean a lot of churn for people updating their projects, as well as might confuse people which work on projects in both editions. I therefore don't expect this to face a lot of acceptance.

  on Linux
- IO operations offered through `libusb`
- IO operations executed in Kernel space (e.g. disk command queues, etc)
- Operations that are composed on top of the described promitives

Async/await is theoretically powerful enough that it allows us to abstract over
such APIs with code like:

```rust
/// Transmit a single byte array using an IO completion based low-level API
async fn transmit_data() {
    let buffer: String = "Data to transfer".into();
    let data = buffer.as_bytes();
    let bytes_transferred = engine.send(data).await;
}
```

This is however not possible with todays `async fn`s. The reason for this is
that the `Future` produced by invoking the function can be dropped while the
execution is still in progress inside `engine.send(data).await`.
If this happens, the engine (which still would hold onto a pointer to the data)
would act on memory that does not represent the actual data to transmit anymore.
This would lead to undefined behavior.

One workaround to this problem is to use owned objects within completion based
operations. E.g.

```rust
async fn transmit_data() {
    let buffer: String = "Data to transfer".into();
    let data: Bytes = buffer.into();

>     let (bytes_transferred, data) = engine.send(data).await;

Reviewer 4: I'm not sure I get how this workaround solves the problem. If `engine.send` is still a multi-step operation you could end up in an inconsistent state by using `abortable`/carelessly using `select` and so on? To me there is no reasonable workaround today (except through docs and manually checking each call site - impossible in e.g. a library). As such, it may give the wrong impression to state that this problem can be worked around.

Author: It mainly solves the problem of corrupting memory if the operation is aborted. With this workaround the memory is temporarily leaked while the operation continues running in the background, but the overall operation is safe.
As you mentioned, it might not solve other logic issues.

amount of bytes into a contiguous buffer using both APIs:

```rust
/// Read `n` bytes using owned buffers into a contiguous buffer
async fn read_n_bytes(n: usize) -> Result<Bytes, IoError> {
    let mut buffer: BytesMut = vec![0; n].into();
    let offset = 0;

    while offset != n {
        let read_buf = buffer.split_at(offset);
        let (bytes_transferred, read_buf) = engine.read(read_buf).await?;
        offset += bytes_transferred;
        buffer = buffer.unsplit(read_buf);
    }

    Ok(buffer.freeze())
}

/// Read `n` bytes using referenced buffers into a contiguous buffer.
/// Note that we could also return an owned buffer as the aggregate result.
/// However in order to showcase the lowest-overhead variant of this method it
/// accepts a borrowed buffer.
async fn read_n_bytes(buffer: &mut [u8]) -> Result<(), IoError> {
    let offset = 0;

    while offset != buffer.len() {
        let bytes_transferred = engine.read(&mut read_buf[offset..]).await?;
        offset += bytes_transferred;
    }

    Ok(())
}
```

## Offer protection against accidental returns

For certain functions it is absolutely necessary for correctness that they
actually runs to completion.

> The reason is typically that the function forms an atomic transaction. If the

Reviewer 4: I understand the buffer use-case is more "real-world", but there's an even simpler example:

```
async fn commit(&mut self) {
  self.a = true;
  something_else().await; // possible exit point
  self.b = true;
}
```

This method can abort at the await point and leave inconsistent data - very different from sync code. The next user of the value gets mutable access, and can't even detect the inconsistent state. Working around this "every await point is possible termination" is infeasible in practice.

Author: Sure. Or just the pattern which is kind of common in other languages:

```
async fn doSth(&mut self) -> Result<(), ()> {
  if self.errored {
    return Err(());
  }
  self.errored = true;
  something_else().await?;
  self.errored = false;
  Ok(())
}
```

There's countless variations of it. The question is which ones are most easy to understand and follow for the audience.

this mechanism is neither performant nor necessarily safe. Blocking on an async
task in a destructor means all other async tasks in the root executor can now
no longer make progress for the duration of the `drop`. And if this mechanism
would be used inside a single-threaded executor a deadlock could occur,
since resources inside `block_on` can require the eventloop which drove the
original `async fn` to make progress - which won't happen since the thread is
blocked on waiting for the cleanup `Future`.

`async fn`s with run to completion semantics will naturally allow transactions
to run to completion - in the absence of panics. Cleanup code can just be regular
code at the end of a function, and no RAII guards are required. Therefore the
new function type will reduce complexity for such transactions.

In addition to that it will lower the chance of accidental errors. There does
not exist an implicit return path in `.await` anymore, which might have not been
handled properly.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The proposal adds 3 new items to Rust:

> - A new `Future` type, which will reside in `core::task`. In the remains of this

Reviewer 4: nit: Trait. This had me confused until the end of this doc, where the trait was explained.

        out.writeln("hello world").await;
        timer.delay(Duration::from_millis(2000)).await;
        out.writeln("hello world").await;
        Ok(())
    }
}
```

Therefore it is important that also a variant of `async` blocks is available
which allow users to `.await` `RunToCompletionFuture`s. The rules around what
`Future` type can be awaited in an `async` block follow the rules for async
functions.

In fact the compiler can implement the validation purely for async blocks,
since every `async fn` will get translated into one of those blocks as
part of the compilation process.

## `#[completion] async fn` support in runtimes

It is expected that runtimes (like [Tokio](https://tokio.rs/) or
[async-std](https://async.rs/)) would either add an additional `spawn` function
or modify their existing `spawn` function to accept a `RunToCompletionFuture`
instead of a `Future`. By adding this support, users can run all kinds of async
functions inside those executors. As long as normal `Future`s can be passed
to methods which accept a `RunToCompletionFuture`, this support can be added
in a backwards compatible fashion. It will not break code already running
on those runtimes.

> ## Cancellation support for run to completion functions

Reviewer 4: Perhaps put in title "graceful cancellation". The way graceful cancellation is implemented with your proposal remains the same as it is today. I've already implemented something similar, for instance. The community would publish and converge on a pattern, perhaps even as part of Context in the future, as you say.

I'd shorten this section and state that graceful cancellation is itself not affected by the proposal, *but* that graceful cancellation becomes more important to get right since futures will increasingly become completion based. This is to avoid confusion that graceful cancellation must be fully resolved in order to make progress - which is not true. We already have the tools to build it!

Author: You are fully right. We can cover graceful cancellation fully separate from this proposal. And in fact I thought about proposing a `stop_token` similar to the C++ one which can be used in sync as well as async contexts.

The main reason this was brought up here is to avoid questions from readers along:
> `Future`s where great because they where so easy to cancel - how do I do the same in this new world?

If you and other feels we should move it out of this doc to focus purely on run to completion, I'm also good with it.

While an `#[completion] async fn` can not be forcefully cancelled - by dropping the
`RunToCompletionFuture` it produced - they can still be cancelled in a cooperative
fashion. Cooperative cancellation consists of 3 phases:
1. Cancellation is signalled to the still running asynchronous function. This can
  e.g. performed by a `CancellationToken` object.
2. The Cancellation requests is detected within the asynchronous function. As a
  result of this cancellation request, the method **can** return to the caller -
  and still deliver a return value. It **could** however also continue to run for
  a certain time. E.g. in order to finalize the important transaction.
3. The issuer of the cancellation request waits for the cancelled async function

>   to return. This can e.g. be achieved through a `WaitGroup` or `Semaphore` type.

Reviewer 4: Not sure I follow why this is necessary? Normally a cancelled function would just return a cancellation error, and the caller just handles the error.

Author: When does the cancellation error get to the caller? The answer here is: When the caller waits for the operation to complete, and notices the error.

So this step is necessary. Note that also in Rust we don't have any standardized exception base class, and therefore also can't have a standardized derived `CancellationError`. This means even if new primitives (e.g. a new file IO API using completion based IO) returns a `CancellationError`, the methods around it might need to translate it to whatever error type the next layer is expecting.

The following example demonstrates cooperative cancellation with the use of a
`CancellationToken`:

```rust
#[completion]
async fn func_with_cancellation_support(
    &mut self, cancel_token: &CancellationToken
) -> Option<i32> {
    let mut last_value = None;
    loop {
        select! {
            val = self.channel.receive() => {
                println!("Received value: {:?}", val);
                last_value = Some(val);
            },

>             _ = cancel_token.cancel_notification().await => {

Reviewer 4: nit: don't think await is needed here.

Author: you are right. Will update when I edit the doc the next time

>             _ = cancel_token.cancel_notification().await => {

Author: ```suggestion
            _ = cancel_token.cancel_notification() => {
```

request while the operation is still pending. The signature of such a method
would e.g. be

```rust
#[completion]
async fn read_with_uring(
    &mut self, buffer: &mut [u8], cancel_token: &CancellationToken
) -> Result<usize, IoError>;
```

The implementation of such a method would rather be complicated - but only needs
to occur within the runtime which provides the method. Most end user code is
simply expected to forward the `CancellationToken`:

```rust
#[completion]
async fn read_all_with_uring(
    reader: &mut Reader, buffer: &mut [u8], cancel_token: &CancellationToken
) -> Result<(), IoError> {
    let mut offset = 0;
    while offset != buffer.len() {
        let read = reader.read_with_uring(&buffer[offset..], cancel_token).await?;
        offset += read;
    }
    Ok(())
}
```

Runtimes could also decide to forward `CancellationToken`s implicitely through
task-local storage. Or the ability to forward a `CancellationToken` could be
added to the [`std::task::Context`](https://doc.rust-lang.org/1.42.0/std/task/struct.Context.html)
type in the future. Such a mechanism is however outside of the scope of this
proposal.

> ## Timeout handling with `#[completion] async fn`

Reviewer 4: Agree on the approach in this section, but this is a special case of graceful cancellation. Perhaps compress or merge with the previous section to make the proposal easier to digest?

behavior for `#[completion] async` blocks compared to `await` blocks:
- Inside `#[completion] async` blocks users are allowed to `.await`
  `RunToCompletionFuture`s and `Future`s.
- Inside `async` blocks users are only allowed to `.await` `Future`s.

## `RunToCompletionFuture` type definition

This RFC proposes to keep `RunToCompletionFuture` closely aligned to the current
`Future` type by defining it in the following fashion:

```rust
pub trait RunToCompletionFuture {
    /// The type of value produced on completion.
    type Output;

    /// Attempt to resolve the future to a final value, registering
    /// the current task for wakeup if the value is not yet available.
    ///
    /// # Return value
    ///
    /// This function returns:
    ///
    /// - [`Poll::Pending`] if the future is not ready yet
    /// - [`Poll::Ready(val)`] with the result `val` of this future if it
    ///   finished successfully.
    ///
    /// If a call to `poll()` call was issued to a `RunToCompletionFuture` that
    /// returned `Pending`, the caller **must** call `poll()` again later,
    /// until it returns `Ready`.
    ///

>     /// Callers are not allowed `drop()` a future which returned `Pending` as

Reviewer 2: ```suggestion
    /// Callers are not allowed to `drop()` a future which returned `Pending` as
```

    /// its last poll result. Futures are only allowed to be dropped if they
    /// either had never been polled, or if the last `poll()` call returned `Ready`.

>     unsafe fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;

Reviewer 4: Is safety really the right boundary to gate this behavior on? We still have panics, so drop-before-ready can still occur. Another example is polling after completion, which is another invariant that the user must respect. Can't we just keep the current signature?

Perhaps a marker trait as a subtrait/supertrait of Future could work? This way, combinators could opt-in to provide the new contract. As proposed, the difference in traits would require anything generic over futures to maintain side-by-side implementations, even if they don't affect "futures are driven to completion" semantics.

Author: If unfortunately seems like the only boundary that the language offers us at this point in time. The issue is that if someone does not poll the function to completion, but the content of the function expects that guarantee - then we would have a memory safety issue.

The example where a sub-task borrows something on the stack of the supertask, but doesn't end within the lifetime of the supertask and gets later resumed is such an example.

Reviewer 2: A supertrait of Future would be fine; any Future is trivially a RunToCompletionFuture.

Panics are a bigger problem.  There are a few solutions I can think of:

- Poll drop: whenever a panic happens in async contexts, any objects on the stack are dropped asynchronously.  The actual panic doesn’t propagate until all the `poll_drop` methods have returned `Poll::Ready`.  I believe this is how `try`-`finally` works in async ECMAScript functions.  Since the only way to call `RunToCompletionFuture::poll` without using `unsafe` is via `.await`, this should work.
- Premature `drop` calls `abort`: *leaking* a polled `RunToCompletionFuture` would still be undefined behavior, but dropping it would call `abort` instead.

Reviewer 4: Thanks for the insights. Our async Rust tasks can run in parallel on multiple threads though, which is quite a different beast.

> The issue is that if someone does not poll the function to completion, but the content of the function expects that guarantee - then we would have a memory safety issue.

Is this implication always true? I buy that the consistency and result of the function is inconsistent and "poisoned", but memory safety sounds like it would only apply in certain cases, such as the subtask-example.

Author: I'm not sure whether there is a big difference between requiring the callers to call `poll_drop` after  a panic, or simply continue to call `poll`. Since everything that the compiler puts into `poll_drop` could be in the same way in `poll`.

Maybe one of the arguments for a dedicated `poll_drop` is that it doesn't have the output value that `poll` has.
I think I would be ok with adding a `poll_drop` to the new `Future` - but more for pure compiler-generated drop glue than for humans to implement.

> Premature drop calls abort: leaking a polled RunToCompletionFuture would still be undefined behavior, but dropping it would call abort instead.

Yeah, that's the "simple way out". For my use-cases I would be rather fine with it, since I think panics + coroutines/futures is a super indeterministic thing anyway. But if others rely on the ability to handle panics in async run to completion contexts, it should indeed be taken into account.

> Is this implication always true? I buy that the consistency and result of the function is inconsistent and "poisoned", but memory safety sounds like it would only apply in certain cases, such as the subtask-example.

It doesn't matter whether its 1 example or thousands. If it's more than 0 examples it has to be labeled as `unsafe` in Rust. People rely on safe functions really being safe - even under heavy API misuse. That's why there also had been very unpleasant discussions for authors of crates which provided support for completion based IO. Those had APIs which very basically always safe, as long as one doesn't do really weird things with them. But still they got push-back from the community about not marking them `unsafe` because theoretically those could lead to memory issues.

do not require run-to-completion semantics - e.g. async Channels, Mutexes and
Semaphores and Timers do very well without them. We might even prefer to cancel
those operations at any time. Only the operations which require run-to-completion
semantics can depend on them - and are still able to use other operations
internally.

Therefore the proposal could also be seens as "providing the best of both worlds":
Users are still able to make use of synchronous cancellation in places where it
is safe and makes sense - which is a feature unique to Rusts `async fn`
implementation. However they will also have to be ability to specify that their
code needs to be run to completion if this is required for correctness of their
application or library.

## Omission of cancellation in the new `RunToCompletionFuture` type

The proposal outlined here does not prescribe how cancellation is performed.
It only mandates a "cancellation signal", and recommends cancelled asynchronous
functions to observe and react to those.

Alternative proposals, like the one in
https://internals.rust-lang.org/t/pre-pre-rfc-unsafe-futures, also proposed to

> directly add a `.cancel()` method on the new `Future` type which needs to initiate

Reviewer 2: As mentioned above, I believe that robust handling of `panic` in `#[completion] async fn`s requires a `.poll_drop()` that is invoked by the compiler in such cases.  This is effectively a `.cancel()` method.

Reviewer 4: Not saying this is a bad idea at all - but cancel and drop are completely different in my mind. Since drop is on the panic path, it simply attempts a best effort cleanup (dropping file descriptors), and protects against memory unsafety.

Cancel OTOH is highly cooperative - and on the expected path. Say you have a search problem where three tasks race to find a solution. Once one is found, the other two are cancelled - gracefully, and awaited as normal (their return values would be Err(Canceled) and discarded).

Reviewer 2: Good point.  I suspect the implementations will often coincide, but that is by no means guaranteed.

mechanism.

`#[completion] async fn` in comparison enforces through the type system that
methods will run to completion, and any necessary cleanup code is run before 
resources are released.

`poll_drop` and `#[completion] async fn` are also similar in the sense that they
both require some `Future`s to be rewritten to support proper cleanup. The
main difference here is that `#[completion] async fn` only requires new `Future`s
that want to support running to completion to be rewritten, and will enforce
this constraint through Rusts type system. `poll_drop` in comparison would require
potentially every existing `Future` and `Future`-combinator to be changed, in
order to provide a similary high guarantee of async cleanup code being performed.

There exists also the possibility to introduce a combination of both proposals:

Since `RunToCompletionFuture` is a new trait, it would be possible to add
`fn poll_drop(self: Pin<&mut Self>)` on this trait, and enforce users to of the
trait to use it: The `unsafe` contract in the trait could require callers to
either `.poll()` the future to completion, or to switch over to calling `poll_drop`.

> However the benefit seems low. By doing this the Future would again move away

Reviewer 2: Panic handling seems to require `poll_drop` or similar, unless one wants to force `panic=abort` for `async` code.

Reviewer 4: I think I agree. The issue I find most hard is during parallelism, i.e. real threads. This occurs in multithreaded async runtimes, but also without async in regular threads. Crossbeam's scoped threads catch the panic and waits for the threads to terminate. Which can wait forever. It's somewhat terrifying that panic propagation is unbounded in time, but similarly terrifying to not be able to statically borrow between threads.. I have no good answers here - just sharing my thoughts.

Reviewer 2: Panic propagation is *already* unbounded in time, since we have Turing-complete destructors.  What this does mean is that we need `std::panic::catch_unwind` to be extremely fast, ideally free.

from one linear code path into an alternate one - which is what run to completion
wants to avoid. It is also highly likely that the code inside `poll_drop` would
actually be equal to the one `poll` itself, since `poll` is also expected to
perform any final cleanup.

> A different alternative might be to pair run to completion methods later with

Reviewer 4: Nice to see, we came to the same conclusion. I think defer is very interesting for multi-exit control flow that we have so much of in Rust with `?`.

some kind of `finally` or `defer` blocks, which allow to run some (potentially
async) code also if methods are exiting early. The benefit of this approach is
that it could work for synchronous methods as well as for async methods.

# Prior art
[prior-art]: #prior-art

Examples for languages which added support for async/await are:
- C#
- Javascript
- C++ (Coroutines)
- Kotlin (Coroutines)
- Zig

In all these languages async functions have run to completion semantics. Rust
is the only language that is known to the author which allows to stop the
execution of an asynchronous function in the middle of the execution. By adopting
this RFC, Rust would gain a mechanism which optionally allows Rusts asynchronous
functions closer to the ones found in other languages, as well as closer to
synchronous functions.

The `CancellationToken/StopToken` approach, which provides the ability to
cooperatively cancel methods with run to completion semantics also had been
successfully deployed in these environments.

> # Unresolved questions

Reviewer 4: Curious about crash consistency, and how this works in the case of panics.

Reviewer 2: If we implement `poll_drop`, then panics could run `poll_drop` to completion as the stack is being unwound.  C++ takes a similar approach via `unhandled_exception`.
